plugins/vpp/l3plugin/descriptor/vrrp.go

```diff
@@ -0,0 +1,235 @@
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+// Delete removes VPP VRRP entry.
+func (d *VrrpDescriptor) Delete(key string, vrrp *l3.VRRPEntry, metadata interface{}) error {
+	if vrrp.Enabled {
+		if err := d.vrrpHandler.VppStopVrrp(vrrp); err != nil {
+			if errors.Is(vppcalls.ErrVRRPUnsupported, err) {
+				d.log.Debugf("Unsupported action: ", err)
+			}
+			return err
+		}
+	}
+	if err := d.vrrpHandler.VppDelVrrp(vrrp); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UpdateWithRecreate returns true if a VRRP update needs to be performed via re-crate.
+func (d *VrrpDescriptor) UpdateWithRecreate(_ string, oldVRRPEntry, newVRRPEntry *l3.VRRPEntry, _ interface{}) bool {
+	if oldVRRPEntry.Enabled == newVRRPEntry.Enabled {
+		// Something has changed, but it is not about enabled/disabled.
+		return true
+	}
+
+	if oldVRRPEntry.Interface == newVRRPEntry.Interface &&
+		oldVRRPEntry.Interval == newVRRPEntry.Interval &&
+		oldVRRPEntry.Priority == newVRRPEntry.Priority &&
+		oldVRRPEntry.VrId == newVRRPEntry.VrId &&
+		oldVRRPEntry.Ipv6 == newVRRPEntry.Ipv6 &&
+		oldVRRPEntry.Accept == newVRRPEntry.Accept &&
+		oldVRRPEntry.Preempt == newVRRPEntry.Preempt &&
+		oldVRRPEntry.Unicast == newVRRPEntry.Unicast &&
+		len(oldVRRPEntry.Addrs) == len(newVRRPEntry.Addrs) {
+		return false
+	}
+
+	for i := 0; i < len(oldVRRPEntry.Addrs); i++ {
+		if oldVRRPEntry.Addrs[i] != newVRRPEntry.Addrs[i] {
+			return true
+		}
+	}
+
+	return true // Something changed except VRRP Enabled = recreate
+}
+
+// Update updates VPP VRRP entry.
+func (d *VrrpDescriptor) Update(_ string, oldVRRPEntry, newVRRPEntry *l3.VRRPEntry, _ interface{}) (
+	_ interface{}, err error) {
+
+	if newVRRPEntry.Enabled {
+		err = d.vrrpHandler.VppStartVrrp(newVRRPEntry)
+	} else {
+		err = d.vrrpHandler.VppStopVrrp(newVRRPEntry)
+	}
+	d.log.Debugf("Unsupported action: ", err)
+
+	return nil, err
+}
+
+// Dependencies lists dependencies for a VPP VRRP entry.
+func (d *VrrpDescriptor) Dependencies(key string, vrrp *l3.VRRPEntry) (deps []kvs.Dependency) {
+	// the interface must exist
+	if vrrp.Interface != "" {
+		deps = append(deps, kvs.Dependency{
+			Label: vrrpEntryInterfaceDep,
+			Key:   interfaces.InterfaceKey(vrrp.Interface),
+		})
+	}
+	return deps
+}
+
+// Retrieve returns all configured VPP VRRP entries.
+func (d *VrrpDescriptor) Retrieve(correlate []adapter.VRRPEntryKVWithMetadata) (
+	retrieved []adapter.VRRPEntryKVWithMetadata, err error,
+) {
+	entries, err := d.vrrpHandler.DumpVrrpEntries()
+
+	for _, entry := range entries {
+
+		retrieved = append(retrieved, adapter.VRRPEntryKVWithMetadata{
+			Key:    l3.VrrpEntryKey(entry.Interface, entry.VrId),
+			Value:  entry,
+			Origin: kvs.UnknownOrigin,
+		})
+	}
+
+	return retrieved, nil
+}
+
+// IsRetriableFailure returns false if error is one of errors
+// defined at the top of this file as non-retriable.
+func (d *VrrpDescriptor) IsRetriableFailure(err error) bool {
+	if errors.Is(err, vppcalls.ErrVRRPUnsupported) {
```

Comment:
You could register this error as global unretrieble error: https://github.com/ligato/vpp-agent/blob/master/plugins/kvscheduler/descriptor_handler.go#L22